lib/identity_cache/cache_fetcher.rb

```diff
@@ -24,27 +75,179 @@ def fetch_multi(keys, &block)
+        data = yield
         break unless IdentityCache.should_fill_cache?
-        result
+        data
+      end
+      data
+    end
+
+    def fetch_with_fill_lock(key, fill_lock_duration, lock_wait_limit)
+      raise ArgumentError, 'fill_lock_duration must be greater than 0.0' unless fill_lock_duration > 0.0
+      raise ArgumentError, 'lock_wait_limit must be greater than 0' unless lock_wait_limit > 0
+      lock = nil
+      using_fallback_key = false
+      expiration_options = EMPTY_HASH
+      (lock_wait_limit + 2).times do # +2 is for first attempt and retry with fallback key
+        result = fetch_or_take_lock(key, old_lock: lock, **expiration_options)
+        case result
+        when FillLock
+          lock = result
+          if lock.client_id == client_id # have lock
+            data = begin
+              yield
+            rescue
+              @cache_backend.cas(key, **expiration_options) do |value|
+                break unless FillLock.cache_value?(value)
+                lock = FillLock.from_cache(value)
+                lock.mark_failed
+                lock.cache_value
+              end
+              raise
+            end
+
+            unless fill_with_lock(key, data, lock, expiration_options)
+              unless using_fallback_key
+                # fallback to storing data in the fallback key so it is available to clients waiting on the lock
+                expiration_options = fallback_key_expiration_options(fill_lock_duration)
+                @cache_backend.write(lock_fill_fallback_key(key, lock), data, **expiration_options)
+              end
+            end
+            return data
+          else
+            raise LockWaitTimeout if lock_wait_limit <= 0
+            lock_wait_limit -= 1
+
+            # If fill failed in the other client, then it might be failing fast
+            # so avoid waiting the typical amount of time for a lock wait. The
+            # semian gem can be used to handle failing fast when the database is slow.
+            if lock.fill_failed?
+              return fetch_without_fill_lock(key) { yield }
+            end
+
+            # lock wait
+            sleep(fill_lock_duration)
+            # loop around to retry fetch_or_take_lock
+          end
+        when IdentityCache::DELETED # interrupted by cache invalidation
+          if lock && !using_fallback_key
+            # cache invalidated during lock wait, try fallback key
+            using_fallback_key = true
+            key = lock_fill_fallback_key(key, lock)
+            expiration_options = fallback_key_expiration_options(fill_lock_duration)
+            # loop around to retry with fallback key
+          else
+            # cache invalidation prevented lock from being taken
+            return yield
+          end
+        when nil # Errors talking to memcached
+          return yield
+        else # hit
+          return result
+        end
+      end
+      raise "unexpected number of loop iterations"
```

Comment:
It seems a bit rough to have a non-specific exception raised here. What can the caller do about it?

I can see two paths that reach here: if the last iteration of the loop was interrupted by cache invalidation (`fetch_or_take_lock` returned `IdentityCache::DELETED` and we have a `lock` and we didn't previously try using the fallback key), or if we `oren`ed on the last iteration of the loop. I feel like we should *not* `raise` here, and we should instead try to handle both those paths in a sane way.

Reply:
> What can the caller do about it?

The caller isn't meant to do anything about it, because getting here means there is a bug in the code.  The `(lock_wait_limit + 2).times do` loop could be replaced with an infinite loop, but I wanted to make it clear that it shouldn't have an infinite loop here, so this check prevents that from happening.

Caching invalidation being interrupted would be the fallback case that is mentioned in the `# +2 is for first attempt and retry with fallback key` comment at the start of the loop.  What does `oren` stand for?  I don't know what you mean by that path.

Reply:
> What does `oren` stand for?

Old `#webscale` that should probably be retired: `sleep`.